beacon-chain/db/kv/state.go

```diff
@@ -264,3 +298,73 @@ func createState(enc []byte) (*pb.BeaconState, error) {
 	}
 	return protoState, nil
 }
+
+// HighestSlotState returns the state with the highest slot from the db.
+func (k *Store) HighestSlotState(ctx context.Context) (*state.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.HighestSlotState")
+	defer span.End()
+	var s *pb.BeaconState
+	err := k.db.View(func(tx *bolt.Tx) error {
+		slotBkt := tx.Bucket(slotsHasObjectBucket)
+		savedSlots := slotBkt.Get(savedStateSlotsKey)
+		highestIndex, err := bytesutil.HighestBitIndex(savedSlots)
+		if err != nil {
+			return err
+		}
+		highestSlot := highestIndex - 1
+		f := filters.NewFilter().SetStartSlot(uint64(highestSlot)).SetEndSlot(uint64(highestSlot))
+
+		keys, err := getBlockRootsByFilter(ctx, tx, f)
+		if err != nil {
+			return err
+		}
+		if len(keys) != 1 {
+			return errors.New("could not get one block root to get state")
+		}
+
+		stateBkt := tx.Bucket(stateBucket)
+		enc := stateBkt.Get(keys[0][:])
+		if enc == nil {
+			return nil
+		}
+
+		s, err = createState(enc)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	if s == nil {
+		return nil, nil
+	}
+
+	return state.InitializeFromProtoUnsafe(s)
+}
+
+// setStateSlotBitField sets the state slot bit in DB.
+// This helps to track which slot has a saved state in db.
+func (k *Store) setStateSlotBitField(ctx context.Context, tx *bolt.Tx, slot uint64) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.setStateSlotBitField")
+	defer span.End()
+
+	bucket := tx.Bucket(slotsHasObjectBucket)
+	slotBitfields := bucket.Get(savedStateSlotsKey)
+	tmp := make([]byte, len(slotBitfields))
+	copy(tmp, slotBitfields)
```

Comment:
You don't need to make a copy here.

Reply:
I had to copy here or I get the following seg fault. It didn't like passing in reference `slotBitfields` directly into `bytesutil.SetBit`. Can you help me understand why that could be a reason? I already spent tons of time on it and didn't have a solution

```
unexpected fault address 0x2f80f6d8
fatal error: fault
[signal SIGBUS: bus error code=0x2 addr=0x2f80f6d8 pc=0x482e870]

goroutine 265 [running]:
runtime.throw(0x4af2796, 0x5)
	GOROOT/src/runtime/panic.go:1112 +0x72 fp=0xc0070ad928 sp=0xc0070ad8f8 pc=0x4036502
runtime.sigpanic()
	GOROOT/src/runtime/signal_unix.go:684 +0x443 fp=0xc0070ad958 sp=0xc0070ad928 pc=0x404d0f3
github.com/prysmaticlabs/prysm/shared/bytesutil.SetBit(...)
	shared/bytesutil/bytes.go:257
github.com/prysmaticlabs/prysm/beacon-chain/db/kv.(*Store).setStateSlotBitField(0xc00701ad50, 0x4c414a0, 0xc00728fc50, 0xc00727b960, 0x0, 0x0, 0x0)
	beacon-chain/db/kv/state.go:371 +0x220 fp=0xc0070ada00 sp=0xc0070ad958 pc=0x482e870
```